util/topsql/reporter/datamodel.go

```diff
@@ -0,0 +1,685 @@
+	// Order by totalCPUTimeMs **DESC**.
+	return rs[i].totalCPUTimeMs > rs[j].totalCPUTimeMs
+}
+
+func (rs records) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+// topN returns the largest n records (by record.totalCPUTimeMs), other
+// records are returned as evicted.
+func (rs records) topN(n int) (top, evicted records) {
+	if len(rs) <= n {
+		return rs, nil
+	}
+	if err := quickselect.QuickSelect(rs, n); err != nil {
+		return rs, nil
+	}
+	return rs[:n], rs[n:]
+}
+
+// toProto converts the records to the corresponding protobuf representation.
+func (rs records) toProto() []tipb.TopSQLRecord {
+	pb := make([]tipb.TopSQLRecord, 0, len(rs))
+	for _, r := range rs {
+		pb = append(pb, r.toProto())
+	}
+	return pb
+}
+
+// collecting includes the collection of data being collected by the reporter.
+type collecting struct {
+	records map[string]*record             // sqlPlanDigest => record
+	evicted map[uint64]map[string]struct{} // { sqlPlanDigest }
+	keyBuf  *bytes.Buffer
+}
+
+func newCollecting() *collecting {
+	return &collecting{
+		records: map[string]*record{},
+		evicted: map[uint64]map[string]struct{}{},
+		keyBuf:  bytes.NewBuffer(make([]byte, 0, 64)),
+	}
+}
+
+// getOrCreateRecord gets the record corresponding to sqlDigest + planDigest, if it
+// does not exist, it will be created.
+func (c *collecting) getOrCreateRecord(sqlDigest, planDigest []byte) *record {
+	key := encodeKey(c.keyBuf, sqlDigest, planDigest)
+	r, ok := c.records[key]
+	if !ok {
+		r = newRecord(sqlDigest, planDigest)
+		c.records[key] = r
+	}
+	return r
+}
+
+// markAsEvicted marks sqlDigest + planDigest under a certain timestamp as "evicted".
+// Later, we can determine whether a certain sqlDigest + planDigest within a certain
+// timestamp has been evicted.
+func (c *collecting) markAsEvicted(timestamp uint64, sqlDigest, planDigest []byte) {
+	if _, ok := c.evicted[timestamp]; !ok {
+		c.evicted[timestamp] = map[string]struct{}{}
+	}
+	c.evicted[timestamp][encodeKey(c.keyBuf, sqlDigest, planDigest)] = struct{}{}
+}
+
+// hasEvicted determines whether a certain sqlDigest + planDigest has been evicted
+// in a certain timestamp.
+func (c *collecting) hasEvicted(timestamp uint64, sqlDigest, planDigest []byte) bool {
+	if digestSet, ok := c.evicted[timestamp]; ok {
+		if _, ok := digestSet[encodeKey(c.keyBuf, sqlDigest, planDigest)]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// appendOthersCPUTime appends totalCPUTimeMs to a special record named "others".
+func (c *collecting) appendOthersCPUTime(timestamp uint64, totalCPUTimeMs uint32) {
+	if totalCPUTimeMs == 0 {
+		return
+	}
+	others, ok := c.records[keyOthers]
+	if !ok {
+		others = newRecord(nil, nil)
+		c.records[keyOthers] = others
+	}
+	others.setCPUTime(timestamp, totalCPUTimeMs)
+}
+
+// appendOthersStmtStatsItem appends stmtstats.StatementStatsItem to a special record named "others".
+func (c *collecting) appendOthersStmtStatsItem(timestamp uint64, item stmtstats.StatementStatsItem) {
```

Comment:
this is also `set` behavior.

Reply:
No need to change the name anymore, because `setStmtStatsItem` have been changed back to `appendStmtStatsItem` (since the implementation has been changed to `merge` semantic).